packages/lisk-transactions/src/transactions/3_vote_transaction.ts

```diff
@@ -0,0 +1,483 @@
+/*
+ * Copyright © 2018 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+import { getAddressAndPublicKeyFromPassphrase } from '@liskhq/lisk-cryptography';
+import { VOTE_FEE } from '../constants';
+import { TransactionError, TransactionMultiError } from '../errors';
+import { Account, Status, TransactionJSON } from '../transaction_types';
+import { prependMinusToPublicKeys, prependPlusToPublicKeys } from '../utils';
+import {
+	isTypedObjectArrayWithKeys,
+	validatePublicKeys,
+	validator,
+} from '../utils/validation';
+import {
+	Attributes,
+	BaseTransaction,
+	createBaseTransaction,
+	CreateBaseTransactionInput,
+	ENTITY_ACCOUNT,
+	EntityMap,
+	RequiredState,
+	TransactionResponse,
+} from './base';
+
+const PREFIX_UPVOTE = '+';
+const PREFIX_UNVOTE = '-';
+const MAX_VOTE_PER_ACCOUNT = 101;
+const MIN_VOTE_PER_TX = 1;
+const MAX_VOTE_PER_TX = 33;
+
+export interface VoteAsset {
+	readonly votes: ReadonlyArray<string>;
+}
+
+export interface CreateVoteAssetInput {
+	readonly unvotes?: ReadonlyArray<string>;
+	readonly votes?: ReadonlyArray<string>;
+}
+
+export type CastVoteInput = CreateBaseTransactionInput & CreateVoteAssetInput;
+
+export interface RequiredVoteState extends RequiredState {
+	readonly dependentState?: {
+		readonly [ENTITY_ACCOUNT]: ReadonlyArray<Account>;
+	};
+}
+
+export const voteAssetTypeSchema = {
+	type: 'object',
+	required: ['votes'],
+	properties: {
+		votes: {
+			type: 'array',
+			items: {
+				type: 'string',
+			},
+		},
+	},
+};
+
+export const voteAssetFormatSchema = {
+	type: 'object',
+	required: ['votes'],
+	properties: {
+		votes: {
+			type: 'array',
+			uniqueSignedPublicKeys: true,
+			minItems: MIN_VOTE_PER_TX,
+			maxItems: MAX_VOTE_PER_TX,
+			items: {
+				type: 'string',
+				format: 'signedPublicKey',
+			},
+		},
+	},
+};
+
+const validateInputs = ({
+	votes = [],
+	unvotes = [],
+}: CreateVoteAssetInput): void => {
+	if (!Array.isArray(votes)) {
+		throw new Error(
+			'Please provide a valid votes value. Expected an array if present.',
+		);
+	}
+	if (!Array.isArray(unvotes)) {
+		throw new Error(
+			'Please provide a valid unvotes value. Expected an array if present.',
+		);
+	}
+	validatePublicKeys([...votes, ...unvotes]);
+};
+
+export class VoteTransaction extends BaseTransaction {
+	public readonly containsUniqueData = true;
+	public readonly asset: VoteAsset;
+
+	public constructor(tx: TransactionJSON) {
+		super(tx);
```

Comment:
Shouldn't `super` call be wrapped in try/catch block? So, we can also validate the asset body to create a complete list of errors in the transaction.

Reply:
it seems like `super` has to be the first sentence, and try breaks the typescript rule...